homeassistant/components/unifi/update.py

```diff
@@ -96,28 +213,40 @@ async def async_added_to_hass(self) -> None:
                 self.async_signal_reachable_callback,
             )
         )
-
-    async def async_will_remove_from_hass(self) -> None:
-        """Disconnect object when removed."""
-        self.controller.entities[DOMAIN][DEVICE_UPDATE].remove(self._obj_id)
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                self.controller.signal_remove,
+                self.remove_item,
+            )
+        )
 
     @callback
     def async_signalling_callback(self, event: ItemEvent, obj_id: str) -> None:
-        """Object has new event."""
-        device = self.controller.api.devices[self._obj_id]
-        self._attr_available = self.controller.available and not device.disabled
-        self._attr_in_progress = device.state == 4
-        self._attr_installed_version = device.version
-        self._attr_latest_version = device.upgrade_to_firmware or device.version
+        """Update the switch state."""
+        if event == ItemEvent.DELETED and obj_id == self._obj_id:
+            self.hass.async_create_task(self.remove_item({self._obj_id}))
+            return
+
+        description = self.entity_description
+        obj = description.object_fn(self.controller.api, self._obj_id)
+        self._attr_available = description.available_fn(self.controller, self._obj_id)
+        self._attr_in_progress = description.state_fn(self.controller.api, obj)
+        self._attr_installed_version = obj.version
+        self._attr_latest_version = obj.upgrade_to_firmware or obj.version
```

Comment:
Most of this code is duplicated in `__init__`
Could it be merged into a shared method?

Reply:
This will be consolidated logic in a common entity class in a few PRs, there might be differences that will be hard to manage in certain platforms so I'd like to keep it like this for now.